storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java

```diff
@@ -0,0 +1,422 @@
+                    _fsOps.deleteIfExists(tr);
+                    _fsOps.deleteIfExists(_stormRoot);
+                }
+            }
+            return null;
+        }
+    }
+    
+    private class DownloadBaseBlobsLocal extends DownloadBaseBlobsDistributed {
+
+        public DownloadBaseBlobsLocal(String topologyId) throws IOException {
+            super(topologyId);
+        }
+        
+        @Override
+        protected void downloadBaseBlobs(File tmproot) throws Exception {
+            _fsOps.forceMkdir(tmproot);
+            String stormCodeKey = ConfigUtils.masterStormCodeKey(_topologyId);
+            String stormConfKey = ConfigUtils.masterStormConfKey(_topologyId);
+            File codePath = new File(ConfigUtils.supervisorStormCodePath(tmproot.getAbsolutePath()));
+            File confPath = new File(ConfigUtils.supervisorStormConfPath(tmproot.getAbsolutePath()));
+            BlobStore blobStore = Utils.getNimbusBlobStore(_conf, null);
+            try {
+                try (OutputStream codeOutStream = _fsOps.getOutputStream(codePath)){
+                    blobStore.readBlobTo(stormCodeKey, codeOutStream, null);
+                }
+                try (OutputStream confOutStream = _fsOps.getOutputStream(confPath)) {
+                    blobStore.readBlobTo(stormConfKey, confOutStream, null);
+                }
+            } finally {
+                blobStore.shutdown();
+            }
+
+            ClassLoader classloader = Thread.currentThread().getContextClassLoader();
+            String resourcesJar = AsyncLocalizer.resourcesJar();
+            URL url = classloader.getResource(ConfigUtils.RESOURCES_SUBDIR);
+
+            String targetDir = tmproot + Utils.FILE_PATH_SEPARATOR + ConfigUtils.RESOURCES_SUBDIR;
+
+            if (resourcesJar != null) {
+                LOG.info("Extracting resources from jar at {} to {}", resourcesJar, targetDir);
+                Utils.extractDirFromJar(resourcesJar, ConfigUtils.RESOURCES_SUBDIR, _stormRoot);
+            } else if (url != null) {
+                LOG.info("Copying resources at {} to {} ", url.toString(), targetDir);
+                if ("jar".equals(url.getProtocol())) {
+                    JarURLConnection urlConnection = (JarURLConnection) url.openConnection();
+                    Utils.extractDirFromJar(urlConnection.getJarFileURL().getFile(), ConfigUtils.RESOURCES_SUBDIR, _stormRoot);
+                } else {
+                    _fsOps.copyDirectory(new File(url.getFile()), new File(targetDir));
+                }
+            }
+        }
+    }
+    
+    private class DownloadBlobs implements Callable<Void> {
+        private final String _topologyId;
+
+        public DownloadBlobs(String topologyId) {
+            _topologyId = topologyId;
+        }
+
+        @Override
+        public Void call() throws Exception {
+            String stormroot = ConfigUtils.supervisorStormDistRoot(_conf, _topologyId);
```

Comment:
The ExecutorService will store the Exception in the Future associated with making the call.  It will be thrown when .get is called on the Future.  We are doing that in Slot.  So in theory we should not miss anything, but yes you are right in practice things can be missed, and it is a huge pain when a bug is introduced and Exceptions are eaten. I'll look into it.